libs/template/renderer.go

```diff
@@ -0,0 +1,302 @@
+		relPath: relPath,
+		content: []byte(content),
+		perm:    perm,
+	}, nil
+}
+
+func (r *renderer) walk() error {
+	directories := []string{"."}
+	var currentDirectory string
+
+	for len(directories) > 0 {
+		currentDirectory, directories = directories[0], directories[1:]
+
+		// Skip current directory if it matches any of accumulated skip patterns
+		instanceDirectory, err := r.executeTemplate(currentDirectory)
+		if err != nil {
+			return err
+		}
+		isSkipped, err := r.isSkipped(instanceDirectory)
+		if err != nil {
+			return err
+		}
+		if isSkipped {
+			logger.Infof(r.ctx, "skipping directory: %s", instanceDirectory)
+			continue
+		}
+
+		// Add skip function, which accumulates skip patterns relative to current
+		// directory
+		r.baseTemplate.Funcs(template.FuncMap{
+			"skip": func(relPattern string) string {
+				// patterns are specified relative to current directory of the file
+				// the {{skip}} function is called from.
+				pattern := path.Join(currentDirectory, relPattern)
+				if !slices.Contains(r.skipPatterns, pattern) {
+					logger.Infof(r.ctx, "adding skip pattern: %s", pattern)
+					r.skipPatterns = append(r.skipPatterns, pattern)
+				}
+				// return empty string will print nothing at function call site
+				// when executing the template
+				return ""
+			},
+		})
+
+		// Process all entries in current directory
+		//
+		// 1. For files: the templates in the file name and content are executed, and
+		//     a in memory representation of the file is generated
+		//
+		// 2. For directories: They are appended to a slice, which acts as a queue
+		//     allowing BFS traversal of the template file tree
+		entries, err := r.templateFiler.ReadDir(r.ctx, currentDirectory)
+		if err != nil {
+			return err
+		}
+		for _, entry := range entries {
+			if entry.IsDir() {
+				// Add to slice, for BFS traversal
+				directories = append(directories, path.Join(currentDirectory, entry.Name()))
+				continue
+			}
+
+			// Generate in memory representation of file
+			f, err := r.computeFile(path.Join(currentDirectory, entry.Name()))
+			if err != nil {
+				return err
+			}
+			logger.Infof(r.ctx, "added file to list of in memory files: %s", f.relPath)
+			r.files = append(r.files, f)
+		}
+
+	}
+	return nil
+}
+
+func (r *renderer) persistToDisk() error {
+	// Accumulate files which we will persist, skipping files whose path matches
+	// any of the skip patterns
+	filesToPersist := make([]*inMemoryFile, 0)
+	for _, file := range r.files {
+		isSkipped, err := r.isSkipped(file.relPath)
+		if err != nil {
+			return err
+		}
+		if isSkipped {
+			log.Infof(r.ctx, "skipping file: %s", file.relPath)
+			continue
+		}
+		filesToPersist = append(filesToPersist, file)
+	}
+
+	// Assert no conflicting files exist
+	for _, file := range filesToPersist {
+		path := file.fullPath()
+		_, err := os.Stat(path)
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to persist to disk, conflict with existing file: %s", path)
```

Comment:
This is tricky, there could be other errors such as a permission issue somewhere in the path.

To check for conflicts, this should check for err == nil (means the path exists), and then a separate branch for all other errors except `os.IsNotExists`, where the returned error wraps it.

Reply:
good catch. done